jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -700,12 +701,97 @@ ecma_op_typedarray_from (ecma_value_t items_val, /**< the source array-like obje
   /* 3 */
   JERRY_ASSERT (ecma_op_is_callable (map_fn_val) || ecma_is_value_undefined (map_fn_val));
 
+  /* 4-5 */
   ecma_object_t *func_object_p = NULL;
   if (!ecma_is_value_undefined (map_fn_val))
   {
     func_object_p = ecma_get_object_from_value (map_fn_val);
   }
 
+  /* 6 */
+  ecma_value_t using_iterator = ecma_op_get_method_by_symbol_id (items_val, LIT_MAGIC_STRING_ITERATOR);
+
+  /* 7 */
+  if (ECMA_IS_VALUE_ERROR (using_iterator))
+  {
+    return using_iterator;
+  }
+
+  ecma_value_t ret_value = ECMA_VALUE_ERROR;
+  ecma_collection_t *values_p = NULL;
+  ecma_value_t new_typedarray;
+  uint32_t len;
+
+  /* 8 */
+  if (!ecma_is_value_undefined (using_iterator))
+  {
+    /* 8.a */
+    ecma_value_t iterator = ecma_op_get_iterator (items_val, using_iterator);
+    ecma_free_value (using_iterator);
+
+    /* 8.b */
+    if (ECMA_IS_VALUE_ERROR (iterator))
+    {
+      return iterator;
+    }
+
+    /* 8.c */
+    values_p = ecma_new_collection ();
+
+    /* 8.e */
+    while (true)
+    {
+      /* 8.e.i */
+      ecma_value_t next = ecma_op_iterator_step (iterator);
+
+      /* 8.e.ii */
+      if (ECMA_IS_VALUE_ERROR (next))
+      {
+        ecma_free_value (iterator);
+        ecma_collection_free (values_p);
+        return next;
+      }
+
+      if (next == ECMA_VALUE_FALSE)
+      {
+        break;
+      }
+
+      /* 8.e.iii */
+      ecma_value_t next_value = ecma_op_iterator_value (next);
+      ecma_free_value (next);
+
+      if (ECMA_IS_VALUE_ERROR (next_value))
+      {
+        ecma_free_value (iterator);
+        ecma_collection_free (values_p);
+        return next_value;
+      }
+
+      ecma_collection_push_back (values_p, next_value);
+    }
+
+    ecma_free_value (iterator);
+
+    /* 8.g */
+    new_typedarray = ecma_typedarray_create_object_with_length (values_p->item_count,
+                                                                NULL,
+                                                                proto_p,
+                                                                element_size_shift,
+                                                                typedarray_id);
+
+    /* 8.h */
+    if (ECMA_IS_VALUE_ERROR (new_typedarray))
+    {
+      ecma_collection_free (values_p);
+      return new_typedarray;
+    }
+
+    len = values_p->item_count;
+
+    goto create_typedarray;
```

Comment:
Do we really need this goto? Can't we just refactor the code a bit? Maybe move some code into methods?

Reply:
I used goto, because in the ECMA standard steps 8.j. and 17. are almost the same, the only difference is when we create the new typedarray from iterable object, we use collections. I can remove the goto and implement the steps from the standard, but that will increase binary size. What is your advice?